src/main/java/org/fungover/thunder/ClientHandler.java

```diff
@@ -17,36 +17,23 @@ public ClientHandler() {
         packageReader = new PackageReader();
     }
 
-    public void handleConnections(ServerSocket serverSocket) throws IOException {
+    public void handleConnections(ServerSocket serverSocket) {
         while (!serverSocket.isClosed()) {
             Socket connection = null;
             try {
                 connection = serverSocket.accept();
+                addNewConnectedClient(connection);
             } catch (IOException e) {
                 e.printStackTrace();
             }
-            if (connection != null) {
-                if (packageReader.isValidConnection(connection)) {
-                    clients.add(connection);
-                }else{
-                    connection.close();
-                }
-            }
             removeDisconnectedClients();
         }
     }
 
-    public void removeDisconnectedClients() {
-        synchronized (clients) {
-            Iterator<Socket> iterator = clients.iterator();
-            while (iterator.hasNext()) {
-                Socket client = iterator.next();
-                if (client.isClosed()) {
-                    System.out.println("Client disconnected: " + client.getInetAddress().getHostName());
```

Comment:
Great job! I think your methods look good. Consider using logger.info(Log4j) instead of System.out.println. It provides better flexibility for handling log messages and is more advantageous in larger projects.

Reply:
good suggestion, i will look into this!

Reply:
If we want to introduce logging using a logging framework like Log4j I suggest that we create an issue for that. Because using a framework means we will have dependencies in runtime that must be included in our Dockerfile so changes must be made there also.

Reply:
good point